File: python/graphs/D860.py
import sys
sys.setrecursionlimit(300000)
input = sys.stdin.readline

n, m = map(int, input().split())

g = [[] for _ in range(n)]
vis = [False] * n
ans = []

for _ in range(m):
    a, b = map(int, input().split())
    a -= 1; b -= 1
    g[a].append(b)
    g[b].append(a)

# true if not used edge, false used edge
def dfs(u:int, f:int) -> bool: 
    vis[u] = True
    curr = []
    for v in g[u]:
        if v != f:
            if vis[v]:
                if u < v:
                    curr.append(v)
            else:
                if dfs(v, u):
                    curr.append(v)

    for i in range(0, len(curr)-1, 2):
        ans.append((curr[i], u, curr[i+1]))

    if len(curr) % 2 == 0:
        return True 

    if f != -1:
        ans.append((curr[-1], u, f))
    return False 

for i in range(n):
    if not vis[i]:
        dfs(i, -1)

print(len(ans))

for a, b, c in ans:
    print(a+1, b+1, c+1)

File: gym/graphs/D860.cpp
#include <iostream>
#include <vector>
using namespace std;
vector<bool> vis;
vector<vector<int>> g;
vector<vector<int>> ans; 

bool dfs(int x, int f) {
    vis[x] = true;
    vector<int> curr;
    for (int y : g[x]) {
        if (y != f) {
            if (vis[y]) {
                // Non-tree edge
                if (x < y) {
                    curr.push_back(y);
                }
            } else {
                // if child return true
                // it did not use edge x-y
                // so x may use it
                if (dfs(y, x)) { 
                    curr.push_back(y);
                }
            }
        }
    }
    for (int i = 0; i+1 < (int)curr.size(); i += 2) {
        ans.push_back({curr[i], x, curr[i+1]});
    }
    if (curr.size() % 2==0) {
        return true; // everyting was paired we dont need egde x-f
    }
    if (f != -1) { // one edge remains 
        ans.push_back({curr.back(), x , f});
    }
    return false; // parent not used
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);

    int n, m; cin >> n >> m;
    g.resize(n);
    vis.resize(n);

    for (int i = 0; i < m; ++i) {
        int u, v; cin >> u >> v;
        u--; v--;
        g[u].push_back(v);
        g[v].push_back(u);
    }

    for (int i = 0; i < n; ++i) {
        if (!vis[i]) dfs(i,-1);
    }

    cout << ans.size() << '\n';
    for (auto& a : ans) {
        cout << a[0]+1 << ' ' << a[1]+1 << ' ' << a[2]+1 << '\n';
    }

    return 0;
}